sdk/core/core/src/az_span.c

```diff
@@ -176,55 +175,43 @@ AZ_NODISCARD az_result az_span_find(az_span source, az_span target, az_span* out
       {
         // All bytes in `target` have been checked and matched the corresponding bytes in `source` (from the start point `i`),
         // so this is indeed an instance of `target` in that position of `source` (step 4.).
-        // Here we create a span in `source` to map the occurrence of `target` in `source` (pardon the redundancy of words, done for explicit clarity).
-        *out_span = az_span_init(source_ptr + i, target_length, target_length);
 
-        return AZ_OK;
+        return i;
       }
     }
   }
 
   // If the function hasn't returned before, all positions 
   // of `source` have been evaluated but `target` could not be found.
-  return AZ_ERROR_ITEM_NOT_FOUND;
+  const int32_t TARGET_NOT_FOUND = -1;
+
+  return TARGET_NOT_FOUND;
 }
 
 AZ_NODISCARD az_span az_span_token(az_span source, az_span delimiter, az_span* out_remainder)
 {
   AZ_PRECONDITION_VALID_SPAN(delimiter, 1, false);
   AZ_PRECONDITION_NOT_NULL(out_remainder);
 
-  if (az_span_is_content_equal(source, AZ_SPAN_NULL))
+  if (az_span_length(source) == 0)
   {
     return AZ_SPAN_NULL;
   }
   else
   {
-    uint8_t* source_ptr = az_span_ptr(source);
-    int32_t source_length = az_span_length(source);
-    int32_t source_capacity = az_span_capacity(source);
+    int32_t index = az_span_find(source, delimiter);
 
-    az_span instance;
-    if (az_span_find(source, delimiter, &instance) == AZ_OK)
+    if (index != -1)
     {
-      uint8_t* instance_ptr = az_span_ptr(instance);
-      int32_t instance_length = az_span_length(instance);
-
-      *out_remainder = az_span_init(
-        instance_ptr + instance_length, 
-        (int32_t)(source_length - instance_length - (instance_ptr - source_ptr)),
-        (int32_t)(source_capacity - instance_length - (instance_ptr - source_ptr)));
-
-      return az_span_init(
-        source_ptr, 
-        (int32_t)(instance_ptr - source_ptr), 
-        (int32_t)(instance_ptr - source_ptr));
+      *out_remainder = az_span_slice(source, index + az_span_length(delimiter), az_span_length(source));
+
+      return az_span_slice(source, 0, index);
     }
     else
     {
       *out_remainder = AZ_SPAN_NULL;
```

Comment:
I'm not sure if the idea of flipping will give the expected functionality we need.

The current design works like this:
The user can keep invoking az_span_token for as long as `source` is not empty (AZ_SPAN_NULL). 
If source is empty, there is not token to be found, so AZ_SPAN_NULL is returned. That is the result that the calling application uses to identify if it should stop calling az_span_token.

On the other hand, if I try to tokenize an array with a sequence that is not present in the array, the whole array is returned as token right now. That is the one point I'm not sure about yet. Maybe it should return empty... the more I'm thinking about it, more I think it should actually return AZ_SPAN_NULL if the delimiter is not in `source`. What do you think, @ahsonkhan ?

Reply:
> If source is empty, there is not token to be found, so AZ_SPAN_NULL is returned. That is the result that the calling application uses to identify if it should stop calling az_span_token.

The same can be done with the remainder instead of with the returned span, so I don't see now that scenario helps us decide one way or another.

> That is the one point I'm not sure about yet. Maybe it should return empty... the more I'm thinking about it, more I think it should actually return AZ_SPAN_NULL if the delimiter is not in `source`.

In either case, one has to be null, and the other has to be the whole source span (either remainder is null, result is the input OR vice versa). I suppose what we pick depends on how you view the "search" happening and what your mental model is of what the remainder means.

I will have to think through a couple of usage patterns, to form an intuition of what I expect the right behavior to be and get back to you (if you have and can share sample code snippets of what we expect callers to do, that would help).

For now, to unblock this PR, I would be fine with either option, so feel free to pick what you think works best, @ewertons. **Leaving it as is works for me too (least code churn when we don't have a strong opinion one way or another).** And we can defer changing it to the future if we find some insight.